goth/runner/probe/agent.py

```diff
@@ -0,0 +1,102 @@
+"""Module for agent components to be used with `Probe` objects."""
+import abc
+import logging
+from typing import Optional, TYPE_CHECKING
+
+from goth.runner.log import LogConfig
+from goth.runner.log_monitor import LogEvent, LogEventMonitor
+from goth.runner.probe.component import ProbeComponent
+
+if TYPE_CHECKING:
+    from goth.runner.probe import Probe
+
+logger = logging.getLogger(__name__)
+
+
+class AgentComponent(ProbeComponent, abc.ABC):
+    """Probe component responsible for managing an agent binary.
+
+    The agent can be an arbitrary binary executed in the container or on host.
+    This class also includes logic for monitoring the agent logs and writing
+    them to a file.
+    """
+
+    log_monitor: LogEventMonitor
+    """Monitor and buffer for agent logs, enables asserting for certain lines to be
+    present in the log buffer.
+    """
+
+    name: str
+    """Name of this agent to be used when creating its log file.
+    This should be unique when compared to other agents running as part of a
+    single test. A good example includes both the probe's name, as well as the
+    type of the agent itself, e.g.: `provider_1_ya-provider`.
+    """
+
+    def __init__(self, probe: "Probe", name: str):
+        super().__init__(probe)
+        self.name = name
+        self._init_log_monitor()
+
+    @abc.abstractmethod
+    async def start(self, *args, **kwargs):
+        """Start the agent binary and initialize the internal log monitor."""
+
+    async def stop(self, *args, **kwargs):
+        """Stop the agent binary and its log monitor."""
+        if self.log_monitor:
+            await self.log_monitor.stop()
+
+    def _init_log_monitor(self):
+        probe = self.probe
+        log_config = LogConfig(file_name=self.name)
+
+        if probe.container.log_config:
+            log_config.base_dir = probe.container.log_config.base_dir
+
+        self.log_monitor = LogEventMonitor(self.name, log_config)
+        self._last_checked_line = -1
+
+    async def wait_for_log(self, pattern: str, timeout: float = 1000) -> LogEvent:
+        """Search agent logs for a log line with the message matching `pattern`."""
+        entry = await self.log_monitor.wait_for_entry(pattern, timeout)
+        return entry
+
+    def __hash__(self):
+        return hash(self.name)
```

Comment:
This seems dangerous, why not use default `__hash__()` here which is based on object identity for custom classes
(https://docs.python.org/3/glossary.html#term-hashable)?

Reply:
This, together with changing `Probe.agents` from a list to a set, was done to prevent adding the same agent type (not necessarily the same class instance) twice to a single `Probe` object (I assumed the same agent type would yield the same name).

We could replace this set + custom hash approach with a dict (that is: make `Probe.agents` a dict in which the keys are names assigned to agents). My problem with this approach is that it makes iterating over agents unnecessarily complex. Also, I think it still makes sense to keep the `name` property on agents, since it's used in `AgentComponent` itself for creating the agent's log file.

Reply:
Why forbid two agent components with different names in a probe? I can imagine a test scenario with two requestor agents running in one node. Sure, they would have different names but the same type.

Moreover, to guarantee no two agent components have the same type I think it would be more natural to define `__hash__()` based on the type, not only the name.

Finally, nothing prevents one from adding two agent components with the same type/name, one of them will be simply silently ignored (probably the second one, but I'm not sure). Maybe instead of overriding `__hash__()` for `AgentComponent`, `Probe` should have a method like this:
```
   def add_agent(agent: AgentComponent):
       if any(a.name == name for a in self.agents):
           raise SomeError(f"The probe already has an agent component named `{agent.name}`")         
      ....
```

Reply:
Thanks for the suggestion, I like this idea. I'll add an `add_agent` method to `Probe`.